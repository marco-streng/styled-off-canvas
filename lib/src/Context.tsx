import * as React from "react";

type Action = { type: "toggle" | "close" };
type Dispatch = (action: Action) => void;
type State = { isOpen: boolean };
type OffCanvasProviderProps = { children: React.ReactNode };

const OffCanvasStateContext = React.createContext<
  { state: State; dispatch: Dispatch } | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      return { isOpen: !state.isOpen };
    }

    case "close": {
      return { isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({ children }: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: false,
  });

  const value = { state, dispatch };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within a OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
