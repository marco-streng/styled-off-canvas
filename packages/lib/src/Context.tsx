import * as React from "react";

type Action = { type: "toggle" | "close" | "open" };

type State = { isOpen: boolean; onClose?: () => void };

type OffCanvasProviderProps = {
  children: React.ReactNode;
  isOpen?: boolean;
  onClose?: () => void;
};

const OffCanvasStateContext = React.createContext<
  | {
      isOpen: boolean;
      onClose?: () => void;
      close: () => void;
      open: () => void;
      toggle: () => void;
    }
  | undefined
>(undefined);

const offCanvasReducer = (state: State, action: Action) => {
  switch (action.type) {
    case "toggle": {
      if (state.isOpen && state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: !state.isOpen };
    }

    case "open": {
      return { ...state, isOpen: true };
    }

    case "close": {
      if (state.onClose) {
        state.onClose();
      }

      return { ...state, isOpen: false };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const OffCanvasProvider = ({
  children,
  isOpen,
  onClose,
}: OffCanvasProviderProps) => {
  const [state, dispatch] = React.useReducer(offCanvasReducer, {
    isOpen: isOpen ?? false,
    onClose,
  });

  const value = {
    isOpen: state.isOpen,
    onClose: state.onClose,
    close: () => dispatch({ type: "close" }),
    open: () => dispatch({ type: "open" }),
    toggle: () => dispatch({ type: "toggle" }),
  };

  return (
    <OffCanvasStateContext.Provider value={value}>
      {children}
    </OffCanvasStateContext.Provider>
  );
};

const useOffCanvas = () => {
  const context = React.useContext(OffCanvasStateContext);
  if (context === undefined) {
    throw new Error("useOffCanvas must be used within an OffCanvasProvider");
  }
  return context;
};

export { OffCanvasProvider, useOffCanvas };
