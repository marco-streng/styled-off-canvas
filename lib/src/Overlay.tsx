import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Div = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $opacity?: number;
  $duration?: string;
}>(
  ({ $isOpen, $background = "#000", $opacity = 0.3, $duration = "500ms" }) => `
  background: ${$background};
  height: 100%;
  left: 0;
  opacity: ${$isOpen ? $opacity : 0};
  position: fixed;
  top: 0;
  transition: opacity ${$duration} ease-in-out;
  width: 100%;
  z-index: ${$isOpen ? 1 : -1};
`
);

/**
 * Overlay component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Overlay = ({
  className,
  background,
  duration,
  opacity,
  ...props
}: {
  background?: string;
  duration?: string;
  opacity?: number;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  return (
    <Div
      className={className}
      onClick={() => close()}
      $isOpen={isOpen}
      $background={background}
      $duration={duration}
      $opacity={opacity}
      {...props}
    />
  );
};
