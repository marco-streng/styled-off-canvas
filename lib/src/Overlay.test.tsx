import { OffCanvasProvider } from "./Context";
import { Overlay } from "./Overlay";
import { render } from "./utils/test-utils";

describe("Overlay", () => {
  test("default", () => {
    const { container } = render(
      <OffCanvasProvider>
        <Overlay />
      </OffCanvasProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
