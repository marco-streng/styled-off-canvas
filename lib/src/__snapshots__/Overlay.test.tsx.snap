// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Overlay > custom color 1`] = `
<div
  class="sc-bcXGCL eOtzel"
/>
`;

exports[`Overlay > custom duration 1`] = `
<div
  class="sc-bcXGCL jnOqTT"
/>
`;

exports[`Overlay > custom opacity 1`] = `
<div
  class="sc-bcXGCL iNTyhs"
/>
`;

exports[`Overlay > default 1`] = `
<div
  class="sc-bcXGCL eXsjSi"
/>
`;

exports[`Overlay > initial open 1`] = `
<div
  class="sc-bcXGCL imPxyi"
/>
`;
